exec/src/main/java/com/palantir/gradle/utils/exec/GradleExec.java

```diff
@@ -37,51 +37,44 @@ public abstract class GradleExec {
-     * to handle exit codes manually.
-     * <p>
-     * The result includes the process's standard output, standard error, and the {@link ExecResult}.
+     * Usage:
+     * <pre>
+     * def result = gradleExec.exec {
+     *     commandLine 'git', 'status'
+     * }
+     *
+     * // Handle success/failure
+     * result.handle(
+     *     { output -> println "Success: ${output.stdOut}" },
+     *     { output -> println "Failed: ${output.stdErr}" }
+     * )
      *
-     * @param action an action to configure the {@link ExecSpec} for the process to be executed
-     * @return a Provider of {@link ExecResultWithOutput} containing the standard output, standard error,
-     *         and execution result
+     * // Or throw on failure
+     * def output = result.get().stdOut
+     * </pre>
      */
-    public Provider<ExecResultWithOutput> exec(Action<? super ExecSpec> action) {
-        Action<ExecSpec> wrappedAction = spec -> {
+    public FailableProvider<ExecResultWithOutput> exec(Action<? super ExecSpec> action) {
+        AtomicReference<String> executable = new AtomicReference<>();
+
+        Action<ExecSpec> captureAction = spec -> {
             action.execute(spec);
+            executable.set(spec.getExecutable());
             spec.setIgnoreExitValue(true);
         };
 
-        ExecOutput execOutput = getProviderFactory().exec(wrappedAction);
-
-        Provider<String> stdoutProvider = execOutput.getStandardOutput().getAsText();
-        Provider<String> stderrProvider = execOutput.getStandardError().getAsText();
-        Provider<ExecResult> resultProvider = execOutput.getResult();
+        ExecOutput execOutput = getProviderFactory().exec(captureAction);
 
-        return getZip().zip3(
-                        resultProvider,
-                        stdoutProvider,
-                        stderrProvider,
+        Provider<ExecResultWithOutput> resultProvider = getZip().zip3(
+                        execOutput.getResult(),
+                        execOutput.getStandardOutput().getAsText(),
+                        execOutput.getStandardError().getAsText(),
                         (result, stdout, stderr) -> ExecResultWithOutput.of(stdout, stderr, result));
-    }
-
-    @Value.Immutable
-    public interface ExecResultWithOutput {
-        String stdOut();
-
-        String stdErr();
-
-        ExecResult result();
 
-        static ExecResultWithOutput of(String stdOut, String stdErr, ExecResult result) {
-            return ImmutableExecResultWithOutput.builder()
-                    .stdOut(stdOut)
-                    .stdErr(stdErr)
-                    .result(result)
-                    .build();
-        }
+        return FailableProvider.of(
+                resultProvider,
+                result -> result.result().getExitValue() != 0,
+                result -> new ExecFailedException(executable.get(), result));
```

Comment:
Hmm, this might introduce more flexibility:

```java
return FailableProvider.of(resultProvider).andThen(
                result ->  result.result().getExitValue() == 0 
                        ?  result
                        :  new ExecFailedException(executable.get(), result));
```